core-to-plc/src/Language/Plutus/Lift.hs

```diff
@@ -0,0 +1,254 @@
+    f <- freshName () "f"
+    return
+        . TyAbs () a (Type ())
+        . Wrap () list (TyApp () listF (TyVar () a))
+        . TyAbs () r (Type ())
+        . LamAbs () z (TyVar () r)
+        . LamAbs () f (TyFun () (TyVar () a) . TyFun () (TyApp () (TyVar () list) (TyVar () a)) $ TyVar () r)
+        $ Var () z
+
+-- |  '(:)' as a term.
+--
+-- > /\(a :: *) -> \(x : a) (xs : list a) ->
+-- >     wrap /\(r :: *) -> \(z : r) (f : a -> list a -> r) -> f x xs
+--
+-- @listA@ appears twice in types in the term,
+-- so this is not really a definition with unique names.
+getBuiltinCons :: Quote (Term TyName Name ())
+getBuiltinCons = do
+    list1 <- getBuiltinList
+    (TyFix () list listF) <- getBuiltinList
+    a  <- freshTyName () "a"
+    x  <- freshName () "x"
+    xs <- freshName () "xs"
+    r  <- freshTyName () "r"
+    z  <- freshName () "z"
+    f  <- freshName () "f"
+    return
+        . TyAbs () a (Type ())
+        . LamAbs () x (TyVar () a)
+        . LamAbs () xs (TyApp () list1 (TyVar () a))
+        . Wrap () list (TyApp () listF (TyVar () a))
+        . TyAbs () r (Type ())
+        . LamAbs () z (TyVar () r)
+        . LamAbs () f (TyFun () (TyVar () a) . TyFun () (TyApp () (TyVar () list) (TyVar () a)) $ TyVar () r)
+        $ mkIterApp (Var () f)
+          [ Var () x
+          , Var () xs
+          ]
+
+-- Generic classes
+
+class GLiftPlc f where
+    gTypeRep :: (MonadQuote m) => m (Type TyName ())
+    gLift :: (MonadQuote m) => f p -> m (Term TyName Name ())
+
+-- Generic instances
+
+{- Note [Specific generic functions]
+We want to define a set of generic functions, but ones which are only valid on certain shapes
+of types. However, we still have to define them via typeclasses. So we end up carefully
+defining a few different functions so that they will only apply to the correct types.
+
+The top-level generic functions `gTypeRep` and `gLift` are only defined for types which
+are metadata-tagged as datatypes. This gives us the root of our function tree. Since we
+don't want it to be callable on other sorts of type, we don't use it recursively but instead
+define new helpers.
+
+`gGetConstructorTypes` gets the types of each constructor, tagged with the name of the constructor.
+Again, we only want to apply this to sums (the ones just inside the datatype declaration), and constructors.
+That means we also need `gGetProductTypes` to handle the product cases inside the constructor, which
+eventually delegates back to `typeRep` for the base case.
+
+The same general pattern applies for `gGetConstructorArgs`.
+-}
+
+-- See Note [Ordering of constructors]
+sortConstructors :: String -> String -> [(String, a)] -> [(String, a)]
+sortConstructors package name cs =
+    let sorted = sortBy (\(n1, _) (n2, _) -> compare n1 n2) cs
+    -- we assume that 'Bool' defined in 'ghc-prim' is the real bool
+    -- TODO: I don't understand why this is 'ghc-prim' rather than 'base', but that's what it seems to be
```

Comment:
Because GHC needs to know about `Bool` (as the type comes with special syntax).

Reply:
Yeah, it's just odd that hackage thinks it's in `base`.